onnxruntime/core/providers/qnn/builder/qnn_backend_manager.cc

```diff
@@ -201,7 +265,247 @@ Status QnnBackendManager::ReleaseContext() {
+    graph_count = binary_info->contextBinaryInfoV1.numGraphs;
+    graphs_info = binary_info->contextBinaryInfoV1.graphs;
+  } else if (binary_info->version == QNN_SYSTEM_CONTEXT_BINARY_INFO_VERSION_2) {
+    graph_count = binary_info->contextBinaryInfoV2.numGraphs;
+    graphs_info = binary_info->contextBinaryInfoV2.graphs;
+  }
+
+  ORT_RETURN_IF(graph_count > 1, "Load from Qnn cached context only support 1 sub-graph.");
+  ORT_RETURN_IF(graphs_info == nullptr, "Failed to get graph info from Qnn cached context.");
+
+  ORT_RETURN_IF(nullptr == qnn_interface_.contextCreateFromBinary,
+                "Invalid function pointer for contextCreateFromBinary.");
+  rt = qnn_interface_.contextCreateFromBinary(backend_handle_,
+                                              device_handle_,
+                                              (const QnnContext_Config_t**)&context_config_,
+                                              static_cast<void*>(buffer.get()),
+                                              buffer_size,
+                                              &context_,
+                                              profile_backend_handle_);
+  ORT_RETURN_IF(QNN_SUCCESS != rt, "Failed to create context from binary.");
+
+  ORT_RETURN_IF_ERROR(qnn_model.DeserializeGraphInfoFromBinaryInfo(graphs_info[0]));
+
+  qnn_sys_interface_.systemContextFree(sys_ctx_handle);
+  sys_ctx_handle = nullptr;
+
+  ORT_RETURN_IF_ERROR(ExtractBackendProfilingInfo());
+  context_created_ = true;
+
+  LOGS(*logger_, VERBOSE) << "Load from cached QNN Context completed.";
+  return Status::OK();
+}
+
+/* \brief: Read string data from binary file with given length
+ * \param[in] binary_file - file stream of the binary file
+ * \param[out] result_str - string read from binary file
+ * \param[out] length - length to read
+ */
+Status ReadStringFromBinaryFile(std::ifstream& binary_file, std::string& result_str, size_t length) {
+  std::unique_ptr<char[]> buffer = std::make_unique<char[]>(length);
+  ORT_RETURN_IF(nullptr == buffer, "Failed to allocate memory for buffer.");
+  const auto& read_result = binary_file.read(buffer.get(), length);
+  ORT_RETURN_IF(!read_result, "Failed to read contents from cached context binary file.");
+  result_str.assign(buffer.get(), length);
+
+  return Status::OK();
+}
+
+/* \brief: Read a uint16_t from binary file
+ * \param[in] binary_file - file stream of the binary file
+ * \param[out] value - uint16_t value
+ */
+Status ReadIntFromBinaryFile(std::ifstream& binary_file, uint16_t& value) {
+  std::unique_ptr<char[]> buffer = std::make_unique<char[]>(sizeof(uint16_t));
+  ORT_RETURN_IF(nullptr == buffer, "Failed to allocate memory for buffer.");
+  const auto& read_result = binary_file.read(buffer.get(), sizeof(uint16_t));
+  ORT_RETURN_IF(!read_result, "Failed to read contents from cached context binary file.");
+  value = *(reinterpret_cast<uint16_t*>(buffer.get()));
```

Comment:
Should able to read a `uint16_t` value without allocating an extra buffer on the heap.

```C++
Status ReadIntFromBinaryFile(std::ifstream& binary_file, uint16_t& value) {
  const auto& read_result = binary_file.read(reinterpret_cast<char*>(&value), sizeof(uint16_t));
  ORT_RETURN_IF(!read_result, "Failed to read contents from cached context binary file.");

  return Status::OK();
}
```

This gets rid of a heap allocation and a potential error case (failed to allocate).